stellargraph/layer/deep_graph_infomax.py

```diff
@@ -0,0 +1,156 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from . import GCN, GAT, APPNP, PPNP
+from ..core.experimental import experimental
+
+from tensorflow.keras.layers import Input, Lambda, Layer, GlobalAveragePooling1D
+import tensorflow as tf
+from tensorflow.keras import backend as K
+import warnings
+
+__all__ = ["DeepGraphInfoMax", "BilinearTransform"]
+
+
+class BilinearTransform(Layer):
+    """
+    This Layer computes the Discriminator function for Deep Graph Infomax (https://arxiv.org/pdf/1809.10341.pdf).
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def build(self, input_shapes):
+
+        self.kernel = self.add_weight(
+            shape=(input_shapes[0][-1], input_shapes[0][-1]),
+            initializer="glorot_uniform",
+            name="kernel",
+            regularizer=None,
+            constraint=None,
+        )
+        self.built = True
+
+    def call(self, inputs):
+        """
+        Applies the layer to the inputs.
+
+        Args:
+            inputs: a list or tuple of tensors with shapes [(N, F), (F,)] containing the node features and summary feature
+                vector.
+        """
+
+        features, summary = inputs
+
+        summary_rank = len(summary.shape)
+        feat_rank = len(features.shape)
+
+        score = tf.tensordot(self.kernel, summary, [[1], [summary_rank - 1]])
+        score = tf.tensordot(features, score, [[feat_rank - 1], [0]])
+
+        return score
+
+
+@experimental(reason="lack of unit tests", issues=[1003])
+class DeepGraphInfoMax:
+    """
+    A class to wrap stellargraph models for Deep Graph Infomax unsupervised training
+    (https://arxiv.org/pdf/1809.10341.pdf).
+
+    Args:
+        base_model: the base stellargraph model class
+    """
+
+    def __init__(self, base_model):
+
+        if not isinstance(base_model, (GCN, GAT, APPNP, PPNP)):
+            raise TypeError(
+                f"base_model: expected GCN, GAT, APPNP or PPNP found {type(base_model).__name__}"
+            )
+
+        self.base_model = base_model
+
+        self._node_feats = None
+
+        # specific to full batch models
+        self._corruptible_inputs_idxs = [0]
+
+    def build(self):
+        """
+        A function to create the the keras inputs and outputs for a Deep Graph Infomax model for unsupervised training.
+
+        Note that the tf.nn.sigmoid_cross_entropy_with_logits loss must be used with this model.
+
+        Example::
+
+            dg_infomax = DeepGraphInfoMax(...)
+            x_in, x_out = dg_infomax.build()
+            model = Model(inputs=x_in, outputs=x_out)
+            model.compile(loss=tf.nn.sigmoid_cross_entropy_with_logits, ...)
+
+        Returns:
+            input and output layers for use with a keras model
+        """
+
+        if self.base_model.multiplicity != 1:
```

Comment:
I think the warning should happen as early as possible, i.e. in `__init__`, so that the user is alerted immediately when they do something incorrect.